src/gnutls.c

```diff
@@ -1049,33 +1050,41 @@ client = INVALID_POOL_HANDLE;
 			tlog (TLOG_COPYCAT, LOG_DEBUG, "Copycat polling returned an error");
 			break;	// Polling sees an error
 		}
-		if (inout [0].revents & POLLIN) {
+		tlog (TLOG_COPYCAT, LOG_DEBUG, "inout [0].revents 0x%04x", inout [0].revents);
+		tlog (TLOG_COPYCAT, LOG_DEBUG, "inout [1].revents 0x%04x", inout [1].revents);
+		if (inout [0].revents & (POLLIN | POLLHUP)) {
 			// Read local and encrypt to remote
-			sz = recv (local, buf, sizeof (buf), RECV_FLAGS);
+			sz = (inout [0].revents & POLLIN) || !local_shutdown_done ? recv (local, buf, sizeof (buf), RECV_FLAGS) : 0;
 			tlog (TLOG_COPYCAT, LOG_DEBUG, "Copycat received %d local bytes (or error<0) from %d", (int) sz, local);
 			if (sz == -1) {
 				tlog (TLOG_COPYCAT, LOG_ERR, "Error while receiving: %s", strerror (errno));
 				break;	// stream error
 			} else if (sz == 0) {
 				inout [0].events &= ~POLLIN;
-				shutdown (local, SHUT_RD);
+				if (!local_shutdown_done) {
+					tlog (TLOG_COPYCAT, LOG_DEBUG, "Sending gnutls_bye");
+					shutdown (local, SHUT_RD);
 #ifdef WINDOWS_PORT
-				setsockopt (remote, SOL_SOCKET, SO_LINGER, (const char *) &linger, sizeof (linger));
+					setsockopt (remote, SOL_SOCKET, SO_LINGER, (const char *) &linger, sizeof (linger));
 #else /* WINDOWS_PORT */
-				setsockopt (remote, SOL_SOCKET, SO_LINGER, &linger, sizeof (linger));
+					setsockopt (remote, SOL_SOCKET, SO_LINGER, &linger, sizeof (linger));
 #endif /* WINDOWS_PORT */
-				gnutls_bye (wrapped, GNUTLS_SHUT_WR);
+					gnutls_bye (wrapped, GNUTLS_SHUT_WR);
+					local_shutdown_done = 1;
+				} else {
+					tlog (TLOG_COPYCAT, LOG_DEBUG, "already done local shutdown");
+				}
 			} else if (gnutls_record_send (wrapped, buf, sz) != sz) {
```

Comment:
hier zou ik dus voorafgaand aan het zenden een voorwaarde toevoegen dat je die kant nog niet afgesloten hebt (negatie is duidelijker: dat die richting nog actief is) *en in deze if..elsif..else is dat zelfs een simpele voorafgaande extra tak die niets doet of enkel iets logt*